src/modules/readout/vmm3/generators/udpgen_vmm3.cpp

```diff
@@ -0,0 +1,98 @@
+///
+/// \file
+///
+/// \brief Generate artificial VMM3 readouts for ESS readout system
+///
+//===----------------------------------------------------------------------===//
+
+#include <CLI/CLI.hpp>
+#include <cinttypes>
+#include <common/Socket.h>
+#include <readout/vmm3/test/ReadoutGenerator.h>
+#include <stdio.h>
+// GCOVR_EXCL_START
+
+struct {
+  /// readout specific
+  uint16_t NRings{2};
+  uint8_t Type{72};            // Freia (see readout ICD for other instruments)
+  /// udp generator generic
+  std::string IpAddress{"127.0.0.1"};
+  uint16_t UDPPort{9000};
+  uint64_t NumberOfPackets{0}; // 0 == all packets
+  uint64_t SpeedThrottle{0};   // 0 is fastest higher is slower
+  uint64_t PktThrottle{0};     // 0 is fastest
+  bool Loop{false};            // Keep looping the same file forever
+
+  bool Randomise{false};       // Randomise header and data
+  // Not yet CLI settings
+  uint32_t KernelTxBufferSize{1000000};
+} Settings;
+
+CLI::App app{"UDP data generator for ESS VMM3 readout"};
+
+int main(int argc, char *argv[]) {
+  app.add_option("-i, --ip", Settings.IpAddress, "Destination IP address");
+  app.add_option("-p, --port", Settings.UDPPort, "Destination UDP port");
+  app.add_option("-a, --packets", Settings.NumberOfPackets,
+                 "Number of packets to send");
+  app.add_option("-t, --throttle", Settings.SpeedThrottle,
+                 "Speed throttle (0 is fastest, larger is slower)");
+  app.add_option("-s, --pkt_throttle", Settings.PktThrottle,
+                 "Extra usleep() after n packets");
+  app.add_option("-y, --type", Settings.Type,
+                 "Detector type id");
+  app.add_option("-r, --rings", Settings.NRings,
+                 "Number of Rings used in data header");
+  app.add_flag("-m, --random", Settings.Randomise, "Randomise header and data fields");
+  app.add_flag("-l, --loop", Settings.Loop, "Run forever");
+
+  CLI11_PARSE(app, argc, argv);
+
+  const int BufferSize{8972};
+  uint8_t Buffer[BufferSize];
+  const uint16_t NumReadouts{400};
+
+  Socket::Endpoint local("0.0.0.0", 0);
+  Socket::Endpoint remote(Settings.IpAddress.c_str(), Settings.UDPPort);
+
+  UDPTransmitter DataSource(local, remote);
+  DataSource.setBufferSizes(Settings.KernelTxBufferSize, 0);
+  DataSource.printBufferSizes();
+
+  uint64_t Packets = 0;
+  uint64_t TotalPackets = 0;
+
+  ReadoutGenerator gen;
+  do {
+    uint32_t SeqNum = TotalPackets;
+    uint16_t DataSize = gen.vmm3ReadoutDataGen(
+          Buffer, BufferSize, Settings.Randomise,
+          Settings.Type, SeqNum, Settings.NRings, NumReadouts);
+
+    DataSource.send(Buffer, DataSize);
+
+    if (Settings.SpeedThrottle) {
+      usleep(Settings.SpeedThrottle);
+    }
+    Packets++;
+    TotalPackets++;
+    if (Settings.PktThrottle) {
+      if (Packets % Settings.PktThrottle == 0) {
+        usleep(10);
+      }
+    }
+    if (Settings.NumberOfPackets != 0 and Packets >= Settings.NumberOfPackets) {
```

Comment:
Maybe I am missing something, but are both `Packets` and `TotalPackets` required? Why is `Packets` set to zero right before breaking the loop, as it is a local variable and not used afterwards?

Reply:
I'll have a look anf fix this. This is copy and paste from udpgen_loki